File: src/xnb/asset/Texture3d.zig
const std = @import("std");

const rh = @import("../reader_helpers.zig");

const PixelFormat = @import("Texture2d.zig").PixelFormat;

const Texture3d = @This();

format: PixelFormat,
width: u32,
height: u32,
depth: u32,
mips: [][]u8,

pub fn initFromReader(reader: *std.Io.Reader, gpa: std.mem.Allocator) !Texture3d {
    const format = try rh.readU32(reader, .little);
    const width = try rh.readU32(reader, .little);
    const height = try rh.readU32(reader, .little);
    const depth = try rh.readU32(reader, .little);
    const mip_count = try rh.readU32(reader, .little);

    var mips = try std.ArrayList([]u8).initCapacity(gpa, mip_count);
    errdefer {
        for (mips.items) |mip| {
            gpa.free(mip);
        }
        mips.deinit(gpa);
    }
    for (0..mip_count) |_| {
        const size = try rh.readU32(reader, .little);
        const mip = try gpa.alloc(u8, size);
        errdefer gpa.free(mip);
        try reader.readSliceAll(mip);
        mips.appendAssumeCapacity(mip);
    }

    return Texture3d{
        .format = @enumFromInt(format),
        .width = width,
        .height = height,
        .depth = depth,
        .mips = try mips.toOwnedSlice(gpa),
    };
}

pub fn deinit(self: *Texture3d, gpa: std.mem.Allocator) void {
    for (self.mips) |mip| {
        gpa.free(mip);
    }
    gpa.free(self.mips);
    self.* = undefined;
}

File: src/xnb/asset/texture_3d.rs
use std::io::Read;

use byteorder::{LittleEndian, ReadBytesExt};

use crate::xnb::asset::texture_2d::PixelFormat;

#[derive(Debug)]
pub struct Texture3D {
    pub format: PixelFormat,
    pub width: u32,
    pub height: u32,
    pub depth: u32,
    pub mips: Vec<Vec<u8>>,
}

impl Texture3D {
    pub fn read(reader: &mut impl Read) -> anyhow::Result<Self> {
        let format = reader.read_u32::<LittleEndian>()?;
        let format = PixelFormat::from_repr(format)
            .ok_or_else(|| anyhow::anyhow!("unknown texture format: {}", format))?;
        let width = reader.read_u32::<LittleEndian>()?;
        let height = reader.read_u32::<LittleEndian>()?;
        let depth = reader.read_u32::<LittleEndian>()?;
        let mip_count = reader.read_u32::<LittleEndian>()?;
        let mut mips = Vec::with_capacity(mip_count as usize);
        for _ in 0..mip_count {
            let size = reader.read_u32::<LittleEndian>()?;
            let mut mip = vec![0u8; size as usize];
            reader.read_exact(&mut mip)?;
            mips.push(mip);
        }
        Ok(Texture3D {
            format,
            width,
            height,
            depth,
            mips,
        })
    }
}
